### PR TITLE
18717 normalize array cache keys

### PR DESCRIPTION
**Pull request by mikk150**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #18717

### CODE REVIEW

File: framework/caching/Cache.php [hunk 1]
```diff
@@ -121,6 +122,31 @@ public function buildKey($key)
         return $this->keyPrefix . $key;
     }
 
+    /**
+     * Normalizes every cache array key element into strings
+     *
+     * @param array $array cache array key
+     * @return array normalized cache array
+     */
+    private function convertArrayValuesToString($array)
```
Comment (samdark) [inline framework/caching/Cache.php:131]: Should be protected since it's used in a pulbic method.

File: framework/caching/Cache.php [hunk 2]
```diff
@@ -109,6 +109,7 @@ public function buildKey($key)
         if (is_string($key)) {
             $key = ctype_alnum($key) && StringHelper::byteLength($key) <= 32 ? $key : md5($key);
         } else {
+            $key = $this->convertArrayValuesToString((array) $key);
```
Comment (bizley) [inline framework/caching/Cache.php:112]: I would wrap it in `if (is_array())` since $key can be whatever.

**Comment by bizley** [comment]
And it requires upgrade info for anyone relying on the old cache behavior.

File: framework/caching/Cache.php (hunk 2 above)
Comment (samdark) [reply framework/caching/Cache.php:112]: It is typecast to array first.

**Comment by mikk150** [comment]
> And it requires upgrade info for anyone relying on the old cache behavior.

if someone relies on the fact that cache keys in an array are strict type

File: framework/caching/Cache.php (hunk 2 above)
Comment (bizley) [reply framework/caching/Cache.php:112]: Yes, I think we should avoid it.

**Comment by bizley** [comment]
The question is, is old `buildKey()` returning always the same values as new `buildKey()`? It's public method.

**Comment by mikk150** [comment]
> The question is, is old `buildKey()` returning always the same values as new `buildKey()`? It's public method.

No, not always
Old system:
`Cache::buildKey(['test', 1]) !== Cache::buildKey(['test', '1'])`
`Cache::buildKey(['test', true]) !== Cache::buildKey(['test', '1'])`
`Cache::buildKey(['test', 1]) !== Cache::buildKey(['test', true])`
`Cache::buildKey(['test', '1']) !== Cache::buildKey(['test', true])`
`Cache::buildKey(1) !== Cache::buildKey('1')`
`Cache::buildKey(true) !== Cache::buildKey('1')`

On new system all those comparisons are true

**Comment by bizley** [comment]
Exactly, I'm just justifying adding upgrade info here.

File: framework/caching/Cache.php (hunk 2 above)
Comment (mikk150) [reply framework/caching/Cache.php:112]: @samdark do you want me to check if key is an array?

Comment (lubosdz) [reply framework/caching/Cache.php:112]: Would something less costy not suffice ? e.g. `$key = md5(var_export($key))`; 
Recursive calculation of a simple cache key seems like quite overhead ... :-)

Comment (mikk150) [reply framework/caching/Cache.php:112]: > Would something less costy not suffice ? e.g. `$key = md5(var_export($key))`;
> Recursive calculation of a simple cache key seems like quite overhead ... :-)

`var_export(1) !== var_export('1')`

Plus it is only recursive if your keys are n+1 dimensional array

Comment (lubosdz) [reply framework/caching/Cache.php:112]: What about `print_r(1, true) === print_r('1', true)` ? Is more [type agnostic](https://3v4l.org/RoG6Y).

Comment (mikk150) [reply framework/caching/Cache.php:112]: if print_r works,(let me test that) then I will probably use that. I would like to benchmark it as well to see which is actually more performant

Comment (mikk150) [reply framework/caching/Cache.php]: so, results are in

`print_r()` won my competition, and wasn't even slower than previous `serialize()`, therefore in my eyes, it is better solution.

My solution was 20% slower

Comment (bizley) [reply framework/caching/Cache.php]: It will be definitely longer. As I think it should work well with objects also, can anyone see any potential problems with object keys?

Comment (mikk150) [reply framework/caching/Cache.php]: no, it will not be longer (storage-wise), because keys are `md5` anyway

Comment (bizley) [reply framework/caching/Cache.php]: Oh, right, true. Question still stands.

Comment (lubosdz) [reply framework/caching/Cache.php]: Waw, I am actually surprised that tests passed, I had doubts about objects. 
I found one [use case to verify](https://3v4l.org/FpGbO) - type juggling (`null - false - 0 - "0"`) - but IMHO all 3 false results are justified, or at least same as current implementation.  
Still I would suggest to test also under `strict_types=1` mode if that does not bring up some unexpected issues.

**Comment by rob006** [comment]
This will treat `null`, `false` and empty string as the same value. I don't think that it is correct - distinction between this three values is pretty common in Yii config and they have different meaning.

**Comment by samdark** [comment]
Isn't that about keys?

**Comment by rob006** [comment]
> Isn't that about keys?

How does it matter? If `false` and `null` have different meaning, they should not generate the same cache key.

**Comment by lubosdz** [comment]
@rob006 
I also pointed that out in the comment above. I would suggest to add more test specifically for these particular cases `null` - `false` - `0` - `""` - `"0"` to make sure there will be no BC break. Currently I don't see any such tests. 
Personally I am also not fan of these "changes" - often seems that fixing one rare case may break other rock-tested stuff.
